lib/spack/spack/solver/asp.py

```diff
@@ -3764,7 +3719,15 @@ def build_specs(self, function_tuples):
         for root in roots.values():
             root._finalize_concretization()
 
-        self._resolve_automatic_splices()
+        # Only attempt to resolve automatic splices if the solver produced any
+        if self._splices:
+            resolved_splices = spack.solver.splicing._resolve_collected_splices(
+                list(self._specs.values()), self._splices
+            )
+            new_specs = {}
+            for node, spec in self._specs.items():
+                new_specs[node] = resolved_splices.get(spec, spec)
+            self._specs = new_specs
```

Comment:
(minor) I wonder if we can change the API slightly, so that here we just have:
```suggestion
            self._specs = spack.solver.splicing.resolve_splices(self._specs, splices=self._splices)
```

Reply:
My goal is to have `Spec.splice` hook into this machinery, hence the desire to leave `NodeArguments` out of the API. It makes does make it more awkward at this call site.

Reply:
Hmm, I think the two goals are not orthogonal. In any case this can be discussed in a following PR.